BJS2-26891 Удаление прошедших событий

Murzin34 opened the pull request.

Murzin34 committed:
BJS2-26891 Удаление прошедших событий

Murzin34 committed:
Небольшой фикс.

Peregruzochka commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -94,6 +99,22 @@ public List<EventDto> getEventParticipants(Long userId) {
+                .stream()
+                .filter(event -> event.getEndDate().isBefore(LocalDateTime.now()))
+                .map(Event::getId)
+                .toList();
```
Я понимаю, что так указано в задании, но я бы достал эти события через sql запрос

Murzin34 replied:
Можно так, можно и так)

Murzin34 replied:
Добавил кастомный `List<Event> findAllByEndDateBefore(LocalDateTime endDate);`

Peregruzochka commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -94,6 +99,22 @@ public List<EventDto> getEventParticipants(Long userId) {
+    }
+
+    @Async("fixedThreadPool")
+    void deleteAllEventsById(List<Long> eventIds) {
```
1. нет модификатора доступа
2. Асинхронные методы должны вызываться из другого компонента. То есть, если мы вызываем метод с аннотацией `@Async` из того же класса, где он определен, то это не будет работать;

Murzin34 replied:
Отказался от использования данного метода и аннотации `@Async`

Murzin34 replied:
Вернул данный метод и аннотацию `@Async`, перенес в EventServiceHelper

Murzin34 committed:
Фикс после кросс-ревью.
Отказался от использования EventExecutorsConfig и аннотации @Async. Поправил нейминг в application.yaml. Поправил тесты.

Murzin34 committed:
Фикс.
Вернул EventExecutorConfig и @Async, метод помеченный @Async перенес в EventServiceHelper, вместо findAll исполузую кастомный findAllByEndDateBefore. Поправил нейминг при имплементации EventServiceHelper в классах EventService, EventServiceTest и EventServiceHelperTest

Murzin34 committed:
Небольшой фикс.

Murzin34 committed:
Небольшой фикс.

Murzin34 committed:
Небольшой фикс.

20cognosce commented on src/main/java/school/faang/user_service/config/event/EventExecutorConfig.java:
```diff
@@ -0,0 +1,20 @@
+public class EventExecutorConfig {
+    private final EventProperties eventProperties;
+
+    @Bean
+    public ExecutorService eventThreadPool() {
+        return Executors.newFixedThreadPool(eventProperties.getThreadsNum());
+    }
```
Для кого он был создан???

Murzin34 replied:
Без него код в сервисе выглядит примерно так: 
```
ListUtils.partition(pastEventsIds, eventProperties.getSublistSize()).forEach(longList -> {
            ExecutorService executor = Executors.newFixedThreadPool(eventProperties.getThreadsNum());

            try {
                executor.submit(() -> eventRepository.deleteAllById(longList));
            } finally {
                executor.shutdown();
                try {
                    if (!executor.awaitTermination(60, TimeUnit.SECONDS)) {
                        executor.shutdownNow();
                    }
                } catch (InterruptedException e) {
                    executor.shutdownNow();
                    Thread.currentThread().interrupt();
                }
            }
        });
```

20cognosce replied:
Добавь квалифаер на него поясняющий что он к ивентам относится или javadoc на этот ListUtils

Murzin34 replied:
Добавил `name = "deletePastEventsThreadPool"` в аннотации `@Bean`

20cognosce commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -94,6 +98,14 @@ public List<EventDto> getEventParticipants(Long userId) {
+    public void deletePastEvents() {
+        List<Long> pastEventsIds = eventRepository.findIdByEndDateBefore(LocalDateTime.now());
+
+        ListUtils.partition(pastEventsIds, eventProperties.getSublistSize())
+                .forEach(eventServiceHelper::deletePastEventsById);
```
Что это за метод? Ты разбиваешь типо айдишники на подсписки, чтобы запустить их асинхронно?
Это не будет работать в транзакции

20cognosce replied:
Каждый подсписок должен иметь свою транзакцию

Murzin34 replied:
Поправил. Высылаю лог в котором удаление 60 событий проходит в рамках 6 транзакций. Также задействуется 4 потока.
[log_Удаление прошедших событий.txt](https://github.com/user-attachments/files/17265601/log_.txt)

Murzin34 committed:
Фикс после ревью.
Поправил реализацию EventService и EventServiceHelper, фикс EventRepository, поправил тесты.

20cognosce commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -94,6 +100,14 @@ public List<EventDto> getEventParticipants(Long userId) {
+        List<Long> pastEventsIds = eventRepository.findIdByEndDateBefore(LocalDateTime.now());
+
+        var sublist = ListUtils.partition(pastEventsIds, eventProperties.getSublistSize());
+        sublist.forEach(eventServiceHelper::deletePastEventsById);
```
Ты не решил потенциальную проблему

Вызов @Async из @Transactional может создать проблемы, потому что каждый @Async работает в отдельном потоке, а @Transactional работает через ThreadLocal + TransactionManager, и каждый вызов deletePastEventsById уже не работает в изначальной транзакции от deletePastEvents, имея другой локальный контекст потока

Наоборот вызывай @Transactional из @Async

Murzin34 replied:
Действительно, косяк. Я почитал https://www.baeldung.com/spring-transactional-async-annotation, попробовал несколько вариантов как задействовать свои методы, в итоге прибился к небольшому изменению того что уже было т.к. поиск евентов и разбивку на подсписки в любом случае нужно делать в рамках метода `@Transactional(readOnly = true)`.

Murzin34 replied:
Высылаю лог файл. 
[log-2024-10-06_0.log](https://github.com/user-attachments/files/17268979/log-2024-10-06_0.log)

20cognosce commented on src/main/java/school/faang/user_service/service/event/EventServiceHelper.java:
```diff
@@ -32,4 +35,10 @@ public void eventExistByIdValidation(Long id) {
     }
+
+    @Async("eventThreadPool")
+    public void deletePastEventsById(List<Long> sublistPastEventsIds) {
```
Этот метод должен быть в транзакции

20cognosce replied:
А он не в транзакции работает

Murzin34 replied:
Повесил `@Transactional` и разбил на 2 метода.

20cognosce commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -94,6 +100,14 @@ public List<EventDto> getEventParticipants(Long userId) {
                 .toList();
     }
 
+    @Transactional
```
Тут ты вызываешь асинхронный метод, в этом потоке ничего не обновляется, поэтому эта транзакция фактически рид онли

Murzin34 replied:
Поправил.

20cognosce reviewed (changes requested):
Покажи старт и коммит транзакций в логах в разрезе потока

Murzin34 committed:
Фикс после ревью.
Поправил реализацию deletePastEvents() в EventService, разбил  удаление sublist`ов в EventServiceHelper, поправил неймин в EventRepository и EventExecutorConfig, поправил тесты. Добавил лог.

Murzin34 committed:
Фикс после ревью.
Убрал лог. Убрал ненужный @Transactional.

20cognosce commented on src/main/java/school/faang/user_service/service/event/EventServiceHelper.java:
```diff
@@ -32,4 +36,16 @@ public void eventExistByIdValidation(Long id) {
         }
     }
+
+    @Transactional
```
Этот Transactional бесполезный
В рамках методов одного бина будет существовать только одна транзакция

20cognosce commented on src/main/java/school/faang/user_service/service/event/EventServiceHelper.java:
```diff
@@ -32,4 +36,16 @@ public void eventExistByIdValidation(Long id) {
+        batchDeletePastEvents(sublistPastEventsIds);
+    }
+
+    @Transactional
```
Тут же уже есть транзакция

20cognosce reviewed (changes requested):
закинь правку через git amend и force push, без нового коммита
